Doc: Add FAQ section for corrupted metadata debugging (#4705)
ArchipelagoMW/Archipelago: docs/apworld_dev_faq.md

We recognize it can feel like a trap since it will not alert you when you are missing an indirect condition, and that some games have very complex access rules.
As of [PR #3682 (Core: Region handling customization)](https://github.com/ArchipelagoMW/Archipelago/pull/3682) being merged, it is possible for a world to opt out of indirect conditions entirely, instead using the system of checking each entrance whenever a region has been reached, although this does come with a performance cost.
Opting out of using indirect conditions should only be used by games that *really* need it. For most games, it should be reasonable to know all entrance &rarr; region dependencies, making indirect conditions preferred because they are much faster.

---

### I uploaded the generated output of my world to the webhost and webhost is erroring on corrupted multidata

> The error `Count not load multidata. File may be corrupted or incompatible.` on uploading a locally generated file is caused because output multidata is the pickled representation of the data that the server needs to know. It may come
> with a descriptive issue like `(No module named 'worlds.myworld')` or `(global 'worlds.myworld.names.ItemNames' is forbidden)`

Reviewer: ```suggestion
The error `Could not load multidata. File may be corrupted or incompatible.` occurs when uploading a locally generated file where there is an issue with the multidata contained within it. It may come
with a descriptive issue like `(No module named 'worlds.myworld')` or `(global 'worlds.myworld.names.ItemNames' is forbidden)`
```

> Pickling is a way to compress python objects as-is in a way that they can be decompressed and be used to rebuild the
> python objects, this means that if you use custom class instances that end up in multidata that the server would need
> to load those custom classes to decompress the data, which can fail either because the world module is unknown or the
> class it's attempting to import to decompress is deemed unsafe.

Reviewer: ```suggestion
Pickling is a way to compress python objects such that they can be decompressed and be used to rebuild the
python objects. This means that if one of your custom class instances ends up in the multidata, the server would not be able
to load that custom class to decompress the data, which can fail either because the custom class is unknown (because it cannot load your world module)
or the class it's attempting to import to decompress is deemed unsafe.
```

> Common situations where this can happen is:
> * Using Option instances directly in slot_data ex. `using options.option_name` instead of `options.option_name.value`,
>   also consider using `options.as_dict("option_name", "option_two")` as a helper
> * Using enums as Location/Item names in the datapackage, when building out location_name_to_id
>   or item_name_to_id make sure to use the underlying str: id mapping

Reviewer: ```suggestion
Common situations where this can happen include:
* Using Option instances directly in slot_data. Ex: using `options.option_name` instead of `options.option_name.value`.
  Also, consider using the `options.as_dict("option_name", "option_two")` helper.
* Using enums as Location/Item names in the datapackage. When building out `location_name_to_id`
  and `item_name_to_id`, make sure the names you are loading in are actually strings and not your enum class.
```

Author: uh, i'm pretty sure we need to call out both the name strings and id ints

Reviewer: Oh, I didn't think people would use an enum to get an int like that, woops

>   and `item_name_to_id`, make sure the names you are loading in are actually strings and not your enum class.

Author: ```suggestion
  and `item_name_to_id`, make sure that you are not using your enum class for either the names or ids in these mappings.
```

We recognize it can feel like a trap since it will not alert you when you are missing an indirect condition, and that some games have very complex access rules.
As of [PR #3682 (Core: Region handling customization)](https://github.com/ArchipelagoMW/Archipelago/pull/3682) being merged, it is possible for a world to opt out of indirect conditions entirely, instead using the system of checking each entrance whenever a region has been reached, although this does come with a performance cost.
Opting out of using indirect conditions should only be used by games that *really* need it. For most games, it should be reasonable to know all entrance &rarr; region dependencies, making indirect conditions preferred because they are much faster.

---

### I uploaded the generated output of my world to the webhost and webhost is erroring on corrupted multidata

The error `Could not load multidata. File may be corrupted or incompatible.` occurs when uploading a locally generated

> file where there is an issue with the multidata contained within it. It may come with a descriptive issue like

Reviewer: ```suggestion
file where there is an issue with the multidata contained within it. It may come with a description like
```
Since "it" is the error, we don't need to mention that the error comes with an issue
